cmd/access/access_node_builder.go

```diff
@@ -193,7 +194,8 @@ func (builder *FlowAccessNodeBuilder) initLibP2PFactory(nodeID flow.Identifier,
 		},
 	}
 
-	libP2PNodeFactory, err := p2p.DefaultLibP2PNodeFactory(builder.Logger,
+	libP2PNodeFactory, err := p2p.DefaultLibP2PNodeFactory(context.Background(),
```

Comment:
Since we're passing in a context to the NodeFactory, we should have some way to cancel it (otherwise it is useless).

We should store the `CancelFunc` (not the context itself) and cancel it in the NodeBuilder's `Done` method.

Reply:
Note that storing the `CancelFunc` is not an antipattern, and is in fact exactly how they were designed to be used.

Reply:
What would be even better yet (but could be deferred to a future PR): The NodeBuilder's `Ready` method creates a single context that is passed down to all of the `Components` and `Modules`. This would probably require a larger refactoring of `scaffold.go` to change [these](https://github.com/onflow/flow-go/blob/master/cmd/scaffold.go#L70-L78) function signatures to take in a context.

Then, instead of calling `Done`, we can simply cancel the context when it is time to shutdown the node.

Reply:
See this discussion for context: https://github.com/onflow/flow-go/pull/1068#discussion_r682830682

I think in the future, we should split the current `ReadyDoneAware` interface into two. One `Runnable` interface with a single method `Run(context.Context)`, which can be used to start a module (and stop it by cancelling the passed in context), and another `ReadyDone` interface with methods `Ready` and `Done` which *only* provide a mechanism for querying the state of the module.

We could implement `Ready` and `Done` with `Run` using `WaitGroups`, ie `Run` would add to the waitgroups for each subcomponent it launches and `Ready` and `Done` would wait for them.

Reply:
Would love feed back on https://github.com/dapperlabs/flow-go/issues/5730 and the attached notion doc :)